Supervised_learning/resnet_lib.py

```diff
@@ -9,28 +10,69 @@
-# Input(filepath: string), Output(tuple of two numpy arrays)
-def create_dataset(filepath):
-    raw_image_dataset = tf.data.TFRecordDataset(filepath)
+IMAGE_SIZE = 64
+EMBEDDING_SIZE = 300
 
+# Choose the correct parsing features based on if there is word embeddings in the dataset.
+# Input(word_embeddings: bool), Output(dictionary)
+def image_feature(word_embeddings):
     # Create a dictionary describing the features.
-    image_feature_description = {
-        'label': tf.io.FixedLenFeature([], tf.int64),
-        'image': tf.io.FixedLenFeature([], tf.string),
-    }
+    if word_embeddings:
+        image_feature_description = {
+            'image': tf.FixedLenFeature([], tf.string),
+            'label': tf.FixedLenFeature([], tf.int64),
+            'texts': tf.io.FixedLenFeature([], tf.string),
+            'embeddings': tf.io.VarLenFeature(dtype=tf.float32)
+        }
+    else:
+        image_feature_description = {
+            'label': tf.io.FixedLenFeature([], tf.int64),
+            'image': tf.io.FixedLenFeature([], tf.string),
+        }
+        
+    return image_feature_description
+
+# Reshpae the embeddings to fit the size of a image channel.
+def reshape_embeddings(embeddings):
+    # EMBEDDING_MAX is the longest embedding size that can fit a channel, for image_size:64 and embeddings_size:300:
+    # the embedding_max is 64*64 // 300 * 300 = 13 * 300 = 3900
+    EMBEDDING_MAX = IMAGE_SIZE**2 // EMBEDDING_SIZE * EMBEDDING_SIZE
+    if tf.shape(embeddings) > EMBEDDING_MAX:
+        embeddings = tf.slice(embeddings, [0], [EMBEDDING_MAX])
+    zero_padding = tf.zeros([IMAGE_SIZE**2] - tf.shape(embeddings), dtype=tf.float32)
+    embeddings_padded = tf.concat([embeddings, zero_padding], 0)
+    result = tf.reshape(embeddings_padded, [IMAGE_SIZE, IMAGE_SIZE, 1])
+    
+    return result
+
 
+# Read in tfrecord files and construct dataset from it
+# Input(filepath: string), Output(tuple of two numpy arrays)
+def create_dataset(filepath, word_embeddings=False):
+    raw_image_dataset = tf.data.TFRecordDataset(filepath)
+    image_feature_description = image_feature(word_embeddings)
+    
     def _parse_image_function(example_proto):
-      # Parse the input tf.Example proto using the dictionary above.
-      return tf.io.parse_single_example(example_proto, image_feature_description)
+        # Parse the input tf.Example proto using the dictionary above.
+        return tf.io.parse_single_example(example_proto, image_feature_description)
 
     parsed_image_dataset = raw_image_dataset.map(_parse_image_function)
     
     image, label = [], []
     
-    for image_features in parsed_image_dataset:
-        label.append(image_features['label'])
-        image.append(tf.image.decode_image(image_features['image']).numpy())
+    if word_embeddings:
+        for image_features in parsed_image_dataset:
+            img = tf.image.decode_image(image_features['image'])
+            img = tf.cast(img, tf.float32) * (2.0 / 255) - 1.0
```

Comment:
Make a comment to explain this.

Reply:
Done

Reply:
Sorry I didn't make it clear. I meant add comments to explain this formula:
`img = tf.cast(img, tf.float32) * (2.0 / 255) - 1.0`

Reply:
Added the comments.